### PR TITLE
Refactor pg_constraint out of DatabaseCatalog.

### PR DESCRIPTION
**Pull request by lmwnshn**
Refactor `pg_constraint` logic away from `DatabaseCatalog` into its own `PgConstraintImpl` object.

Same deal as #1345 which should be merged first.

Just opening it to keep track of merge order.

Note that this does NOT contain the changes from #1077. That will come in a future PR, where we decide if modularizing things is good enough to add that code in, or if we need to go for the DSL.

### CODE REVIEW

File: src/include/catalog/postgres/pg_constraint_impl.h [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+  /**
+   * Bootstrap:
+   *    pg_constraint
+   *    pg_constraint_oid_index
+   *    pg_constraint_name_index
+   *    pg_constraint_namespace_index
+   *    pg_constraint_table_index
+   *    pg_constraint_index_index
+   *    pg_constraint_foreigntable_index
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_constraint_impl.h:51]: Add dependency info.

File: src/include/catalog/postgres/pg_constraint_impl.h [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * Prepare to create pg_constraint.
+   * Does NOT create anything until the relevant bootstrap functions are called.
+   *
+   * @param db_oid          The OID of the database that pg_constraint should be created in.
+   */
+  explicit PgConstraintImpl(db_oid_t db_oid);
+
+  /** Bootstrap the projected row initializers for pg_constraint. */
+  void BootstrapPRIs();
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_constraint_impl.h:41]: Make private and call from `Bootstrap`

File: src/include/catalog/postgres/pg_constraint_impl.h [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+   * Get all of the expressions used for constraints, used in TearDown.
+   * A buffer is passed in to allow for buffer reuse optimizations.
+   *
+   * @param txn             The transaction to obtain all the expressions in.
+   * @param buffer          The buffer to use for getting function contexts back out.
+   * @param buffer_len      The length of buffer. Only used in debug mode to assert correctness.
+   * @return All the expressions used for constraints.
+   */
+  std::vector<parser::AbstractExpression *> TearDownGetExpressions(
+      common::ManagedPointer<transaction::TransactionContext> txn, byte *buffer, UNUSED_ATTRIBUTE uint64_t buffer_len);
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_constraint_impl.h:69]: See question in #1344 of whether this should return an opaque function that encapsulates the deletions instead.

File: src/include/catalog/postgres/pg_constraint_impl.h [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  const db_oid_t db_oid_;
+
+  storage::SqlTable *constraints_;
+  storage::index::Index *constraints_oid_index_;
+  storage::index::Index *constraints_name_index_;  // indexed on namespace OID and name
+  storage::index::Index *constraints_namespace_index_;
+  storage::index::Index *constraints_table_index_;
+  storage::index::Index *constraints_index_index_;
+  storage::index::Index *constraints_foreigntable_index_;
+};
```
Comment (jrolli) [inline src/include/catalog/postgres/pg_constraint_impl.h:84]: Where are the pre-generated projected row initializers?

**Review by jrolli — CHANGES_REQUESTED** [review]
Pretty much same comments as #1344 and #1345.

File: src/include/catalog/postgres/pg_constraint_impl.h (hunk 1 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_constraint_impl.h:51]: wip

File: src/include/catalog/postgres/pg_constraint_impl.h (hunk 2 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_constraint_impl.h:41]: Per previous conversation, needs builder refactor

File: src/include/catalog/postgres/pg_constraint_impl.h (hunk 3 above)
Comment (lmwnshn) [reply src/include/catalog/postgres/pg_constraint_impl.h]: Refactored, good idea

**Comment by lmwnshn** [comment]
Will merge with #1358.

**Pull request closed without merging**